### ISSUE TITLE
There may be a bug in function proxy_writer_add_connection.

### ISSUE THREAD

**Issue by tooyoungtoosimpo** [eclipse-cyclonedds/cyclonedds#1874]
if (ddsi_is_builtin_entityid (rd->e.guid.entityid, NN_VENDORID_ECLIPSE) && !ddsrt_avl_is_empty (&pwr->readers) && !pwr->filtered)
  {
    /* builtins really don't care about multiple copies or anything */
    m->in_sync = PRMSS_SYNC;
  }
  
  There may have a bug while deleting the first matched reader.
  Think about this case:
  While process A create a participant(dp1) and a writer, process B have a participant(dp2) before this, and then process B create a new participant(dp3).
  Now the interactive process for BUILTIN_PUBLICATIONS_WRITER will be like:
![微信图片_20231109180256](https://github.com/eclipse-cyclonedds/cyclonedds/assets/103105551/91ba7b0f-ff3b-4109-ad61-def244f38b4d)
In the process above, after step1, the in_sync for dp2's BUILTIN_PUBLICATIONS_READER and dp1's BUILTIN_PUBLICATIONS_WRITER is PRMSS_OUT_OF_SYNC. And if we create a new participant(dp3) between step1 and step2, the in_sync for dp3's BUILTIN_PUBLICATIONS_READER and dp1's BUILTIN_PUBLICATIONS_WRITER will be PRMSS_SYNC because of the judgement " !ddsrt_avl_is_empty (&pwr->readers)" is true.
But if we delete the dp2 between step1 and step2, the data(w) of dp1 will never be sent to dp3.

**Comment by eboasson** [comment]
This is the type of issue I like best 😀

Short answer: I believe you're right.

Long answer: perhaps you're curious where this came from. It comes from OpenSplice where it does wonders for reducing the work involved when there are many participants handled by a single DDSI service. This is quite a common case for configurations using shared memory: hundreds of processes/participants and 10k+ readers/writers per node are totally normal, and this DDSI stack handles that gracefully because of such tricks.

If you look at https://github.com/ZettaScaleLabs/ospli/blob/3af164c18c87cf6e958fe50ad9a23fc6a1469a77/src/services/ddsi2/core/q_entity.c#L2299 you can see the striking similarity 🙂 But in OpenSplice your `dp2` is never deleted because it is associated with the service handling DDSI traffic. There is a related thing: the "minimal built-in endpoint set mode" which you can use in OpenSplice but not (yet) in Cyclone because it uses only one set of discovery *writers* and so absolutely needs such a "forever" participant. It is a nice idea, it really reduces discovery overhead when you have multiple participants.[^1] I've kept the support code because it can still work if one simply creates that "forever" participant if the mode is enabled.

I have had a nagging feeling I should look into this in Cyclone but it got postponed because, well, because the usual reasons and because the pattern seems to be so rare that I could get away with postponing looking into it.

So thank you! ☺️ Now let's see what the best solution is ...

[^1]: This is really a more refined version of the old "squash participants" mode that I put in OpenSplice way back when and that has come in useful many times, but which cannot work at all in Cyclone despite the existence of the option: https://github.com/eclipse-cyclonedds/cyclonedds/issues/1811.

**Comment by tooyoungtoosimpo** [comment]
This problem cause a bug in my program.
My temporary solution is :
`if (ddsi_is_builtin_entityid (rd->e.guid.entityid, NN_VENDORID_ECLIPSE) && !pwr->filtered)
  {
    /* builtins really don't care about multiple copies or anything */ //warning: care!!! very care!!!
    m->in_sync = PRMSS_OUT_OF_SYNC;
    m->u.not_in_sync.end_of_tl_seq = MAX_SEQ_NUMBER;
  }`
This solution may generate a lot of unnecessary extra data(w) or data(r) while creating new participants. But fortunately, I don't have a large number of writers or readers in my program.

I hope to see your solution soon.😀
